File: librarian/scripts/hawp/work-validate/__tests__/validations.test.ts
import assert from "node:assert/strict";
import { mkdtempSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import test from "node:test";

import { checkBacklogConsistency } from "../validations/backlog-consistency";
import { checkClosedTaskCompleteness } from "../validations/closed-task-completeness";
import {
  checkEvidenceIntegrity,
  collectClosedPlanFiles,
} from "../validations/evidence-integrity";
import { checkVerificationClarity } from "../validations/verification-clarity";

const withTempWorkDir = (fn: (workDir: string) => void): void => {
  const root = mkdtempSync(join(tmpdir(), "hawp-validations-"));
  try {
    const workDir = join(root, ".hawp", "work");
    mkdirSync(join(workDir, "active"), { recursive: true });
    mkdirSync(join(workDir, "parked"), { recursive: true });
    mkdirSync(join(workDir, "closed"), { recursive: true });
    fn(workDir);
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
};

const writeClosedPlan = (
  workDir: string,
  date: { year: string; month: string; day: string },
  fileName: string,
  content: string,
): string => {
  const dir = join(workDir, "closed", date.year, date.month, date.day);
  mkdirSync(dir, { recursive: true });
  const filePath = join(dir, fileName);
  writeFileSync(filePath, content, "utf-8");
  return filePath;
};

test("checkVerificationClarity accepts plain and annotated Verification headings", () => {
  withTempWorkDir((workDir) => {
    const plain = writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-900.md",
      "# Task\n\n## Verification\n\n- [x] checked. Evidence: output captured\n",
    );
    const annotated = writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-901.md",
      "# Task\n\n## Verification (filled at close)\n\n- [ ] NOT YET VERIFIED claim\n",
    );

    const result = checkVerificationClarity([plain, annotated]);
    assert.equal(result.total, 2);
    assert.equal(result.proven, 1);
    assert.equal(result.unproven.length, 1);
    assert.equal(result.ambiguous.length, 0);
    assert.equal(typeof result.unproven[0]?.lineNumber, "number");
    assert.match(result.unproven[0]?.filePath ?? "", /TASK-901\.md$/);
    assert.equal(result.status, "WARN");
  });
});

test("checkVerificationClarity warns on checklist claims without evidence labels", () => {
  withTempWorkDir((workDir) => {
    const ambiguous = writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "02" },
      "TASK-902.md",
      "# Task\n\n## Verification\n\n- [x] Smoke test completed\n",
    );

    const result = checkVerificationClarity([ambiguous]);
    assert.equal(result.total, 1);
    assert.equal(result.proven, 0);
    assert.equal(result.unproven.length, 0);
    assert.equal(result.ambiguous.length, 1);
    assert.equal(result.ambiguous[0]?.id, "TASK-902");
    assert.equal(result.ambiguous[0]?.lineNumber, 2);
    assert.match(result.ambiguous[0]?.filePath ?? "", /TASK-902\.md$/);
    assert.equal(result.status, "WARN");
  });
});

test("checkVerificationClarity ignores evidence follow-up scaffolding entries", () => {
  withTempWorkDir((workDir) => {
    const plan = writeClosedPlan(
      workDir,
      { year: "2026", month: "07", day: "09" },
      "TASK-903.md",
      [
        "# Task",
        "",
        "## Verification",
        "",
        "- [x] Smoke test completed. Evidence: command output recorded",
        "",
        "### Evidence Follow-Up",
        "",
        "- [ ] Research evidence for: Old claim",
        "- [ ] Update the original verification checklist line with Evidence: ... or explicit unproven wording.",
        "",
      ].join("\n"),
    );

    const result = checkVerificationClarity([plan]);
    assert.equal(result.total, 1);
    assert.equal(result.proven, 1);
    assert.equal(result.unproven.length, 0);
    assert.equal(result.ambiguous.length, 0);
    assert.equal(result.status, "PASS");
  });
});

test("checkEvidenceIntegrity validates links and rejects evidence-folder escapes", () => {
  withTempWorkDir((workDir) => {
    const evidenceDir = join(workDir, "evidence", "2026", "06", "01");
    mkdirSync(evidenceDir, { recursive: true });
    writeFileSync(join(evidenceDir, "TASK-900-proof.md"), "proof", "utf-8");

    const plan = writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-900.md",
      [
        "# Task",
        "",
        "## Verification",
        "",
        "- Evidence: link to ../evidence/2026/06/01/TASK-900-proof.md",
        "- Evidence: link to ../evidence/2026/06/01/missing.md",
        "- Evidence: link to ../evidence/../../escape.md",
        "",
      ].join("\n"),
    );

    const result = checkEvidenceIntegrity(workDir, [plan]);
    assert.equal(result.total, 2, "escaping link must not be counted");
    assert.equal(result.valid, 1);
    assert.equal(result.broken.length, 1);
    assert.equal(result.status, "WARN");
  });
});

test("collectClosedPlanFiles walks only YYYY/MM/DD folders", () => {
  withTempWorkDir((workDir) => {
    writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-900.md",
      "# Task\n",
    );
    writeFileSync(join(workDir, "closed", "README.md"), "readme", "utf-8");

    const files = collectClosedPlanFiles(join(workDir, "closed"));
    assert.equal(files.length, 1);
    assert.ok(files[0]?.endsWith("TASK-900.md"));
  });
});

test("checkBacklogConsistency passes when rows match files and fails on missing plans", () => {
  withTempWorkDir((workDir) => {
    writeFileSync(
      join(workDir, "active", "TASK-100.md"),
      "# plan\n",
      "utf-8",
    );

    const okResult = checkBacklogConsistency(workDir, {
      active: [
        {
          id: "TASK-100",
          type: "task",
          title: "sample",
          status: "in-progress",
          detail: "[plan](active/TASK-100.md)",
        },
      ],
      closed: [],
      parked: [],
    });
    assert.equal(okResult.status, "PASS");
    assert.equal(okResult.activeWork.found, 1);

    const missingResult = checkBacklogConsistency(workDir, {
      active: [
        {
          id: "TASK-200",
          type: "task",
          title: "missing plan",
          status: "in-progress",
          detail: undefined,
        },
      ],
      closed: [],
      parked: [],
    });
    assert.equal(missingResult.status, "FAIL");
    assert.deepEqual(missingResult.activeWork.missing, ["TASK-200"]);
  });
});

test("checkBacklogConsistency matches short-UUID rows to full-UUID files without orphans", () => {
  withTempWorkDir((workDir) => {
    const uuid = "0e1c4afa-9668-4d61-b5b6-1e27be42ca23";
    writeFileSync(join(workDir, "active", `${uuid}.md`), "# plan\n", "utf-8");

    const result = checkBacklogConsistency(workDir, {
      active: [
        {
          id: "0e1c4afa",
          type: "improvement",
          title: "short-uuid display row",
          status: "in-progress",
          detail: `[plan](active/${uuid}.md)`,
        },
      ],
      closed: [],
      parked: [],
    });
    assert.equal(result.status, "PASS");
    assert.equal(result.activeWork.found, 1);
    assert.deepEqual(result.orphanedFiles, []);
  });
});

test("checkBacklogConsistency resolves UUID-named active plan files without orphans", () => {
  withTempWorkDir((workDir) => {
    const uuid = "361fb08e-6457-4ed5-80bd-76337b6f0e89";
    writeFileSync(join(workDir, "active", `${uuid}.md`), "# plan\n", "utf-8");

    const result = checkBacklogConsistency(workDir, {
      active: [
        {
          id: uuid,
          type: "task",
          title: "uuid-native item",
          status: "in-progress",
          detail: `[plan](active/${uuid}.md)`,
        },
      ],
      closed: [],
      parked: [],
    });
    assert.equal(result.status, "PASS");
    assert.equal(result.activeWork.found, 1);
    assert.deepEqual(result.orphanedFiles, []);
  });
});

test("checkClosedTaskCompleteness recognizes a shared closed record via its Closes: line", () => {
  withTempWorkDir((workDir) => {
    writeClosedPlan(
      workDir,
      { year: "2026", month: "07", day: "26" },
      "shared-audit.md",
      [
        "# Shared audit",
        "",
        "**Closes:** `abc12345`, `def67890`",
        "",
        "## Outcome",
        "",
        "Both fixed.",
        "",
        "## Verification",
        "",
        "- [x] checked. Evidence: output captured",
        "",
        "## Close Checklist",
        "",
        "- [x] done",
        "",
      ].join("\n"),
    );

    const result = checkClosedTaskCompleteness(workDir, []);
    assert.equal(result.total, 1);
    assert.equal(result.untypedCurrent.length, 0);
    assert.equal(result.failing.length, 0);
    assert.equal(result.status, "PASS");
  });
});

test("checkBacklogConsistency matches multiple rows against one shared closed record", () => {
  withTempWorkDir((workDir) => {
    writeClosedPlan(
      workDir,
      { year: "2026", month: "07", day: "26" },
      "shared-audit.md",
      "# Shared audit\n\n**Closes:** `abc12345`, `def67890`\n",
    );

    const result = checkBacklogConsistency(workDir, {
      active: [],
      closed: [
        {
          id: "abc12345",
          type: "fix",
          title: "first fix",
          status: "done",
          detail: undefined,
        },
        {
          id: "def67890",
          type: "fix",
          title: "second fix",
          status: "done",
          detail: undefined,
        },
      ],
      parked: [],
    });
    assert.equal(result.status, "PASS");
    assert.equal(result.recentlyClosed.found, 2);
  });
});

test("checkClosedTaskCompleteness fails post-cutoff plans missing required sections", () => {
  withTempWorkDir((workDir) => {
    writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-910.md",
      "# Task\n\n## Outcome\n\nDone.\n\n## Verification\n\n- [x] ok\n\n## Close Checklist\n\n- [x] done\n",
    );
    writeClosedPlan(
      workDir,
      { year: "2026", month: "06", day: "01" },
      "TASK-911.md",
      "# Task\n\nNo close sections here.\n",
    );

    const result = checkClosedTaskCompleteness(workDir, []);
    assert.equal(result.total, 2);
    assert.equal(result.failing.length, 1);
    assert.equal(result.failing[0]?.id, "TASK-911");
    assert.equal(result.status, "FAIL");
  });
});

File: librarian/go/internal/domain/work/validations_test.go
package work

import (
	"path/filepath"
	"testing"

	"github.com/sentzunhat/hawp/librarian/go/internal/domain/work/source"
)

const closedPlanComplete = `# done thing

## Outcome (filled at close)

Done.

## Verification (filled at close)

- [x] It works (Evidence: test run output)

## Close Checklist

- [x] Outcome section filled
`

// TestBacklogConsistencyPassAndFail tests the snapshot-based consistency check
// without touching the filesystem (non-repository path).
func TestBacklogConsistencyPassAndFail(t *testing.T) {
	snapshot := source.Snapshot{
		Files: []source.File{
			{RelPath: "active/39bc92b6-5e3f-49ae-970d-6126ec2cfd82.md"},
			{RelPath: "closed/2026/07/03/TASK-086.md"},
			{RelPath: "parked/bee15107.md"},
		},
	}
	backlog := &Backlog{
		Active: []BacklogRow{{ID: "39bc92b6"}}, // short-UUID row → full-UUID file
		Closed: []BacklogRow{{ID: "TASK-086"}},
		Parked: []BacklogRow{{ID: "bee15107", Detail: "[plan](parked/bee15107.md)"}},
	}
	result := CheckBacklogConsistency(snapshot, backlog)
	if result.Status != StatusPass {
		t.Fatalf("status = %s, want PASS: %+v", result.Status, result)
	}

	// A row without a matching file is reported as missing.
	backlog.Active = append(backlog.Active, BacklogRow{ID: "deadbeef"})
	result = CheckBacklogConsistency(snapshot, backlog)
	if result.Status != StatusFail || len(result.ActiveWork.Missing) != 1 {
		t.Fatalf("missing row not detected: %+v", result.ActiveWork)
	}

	// An active plan file whose ID is not in the backlog is reported as orphaned.
	orphanSnapshot := source.Snapshot{
		Files: []source.File{
			{RelPath: "active/TASK-999.md"},
		},
	}
	result = CheckBacklogConsistency(orphanSnapshot, &Backlog{})
	if result.Status != StatusFail || len(result.OrphanedFiles) != 1 {
		t.Fatalf("orphan not detected: %+v", result.OrphanedFiles)
	}
}

func TestClosedTaskCompleteness(t *testing.T) {
	snapshot := source.Snapshot{Files: []source.File{
		{Path: "/work/closed/2026/07/03/TASK-086.md", RelPath: "closed/2026/07/03/TASK-086.md", Content: closedPlanComplete},
		{Path: "/work/closed/2026/07/03/TASK-087.md", RelPath: "closed/2026/07/03/TASK-087.md", Content: "# plan without sections"},
		{Path: "/work/closed/2026/04/01/TASK-001.md", RelPath: "closed/2026/04/01/TASK-001.md", Content: "# legacy plan without sections"},
		{Path: "/work/closed/2026/07/03/TASK-086-summary.md", RelPath: "closed/2026/07/03/TASK-086-summary.md", Content: "# supporting, skipped"},
	}}
	result := CheckClosedTaskCompleteness(snapshot)
	if result.Total != 3 {
		t.Fatalf("total plans = %d, want 3", result.Total)
	}
	if result.Skipped != 1 {
		t.Errorf("skipped = %d, want 1 (summary suffix)", result.Skipped)
	}
	if len(result.Failing) != 1 || result.Failing[0].ID != "TASK-087" {
		t.Errorf("failing = %+v, want TASK-087", result.Failing)
	}
	if len(result.Warnings) != 1 || result.Warnings[0].ID != "TASK-001" {
		t.Errorf("warnings = %+v, want legacy TASK-001", result.Warnings)
	}
	if result.Status != StatusFail {
		t.Errorf("status = %s, want FAIL", result.Status)
	}

	if got := CheckClosedTaskCompleteness(source.Snapshot{Files: []source.File{
		{Path: "/work/closed/2026/07/03/TASK-086.md", RelPath: "closed/2026/07/03/TASK-086.md", Content: closedPlanComplete},
	}}); got.Status != StatusPass {
		t.Errorf("all-complete status = %s, want PASS", got.Status)
	}
}

// TestEvidenceIntegrity tests the snapshot-based evidence check without
// touching the filesystem (non-repository path).
func TestEvidenceIntegrity(t *testing.T) {
	snapshot := source.Snapshot{
		ClosedFiles: []source.File{
			{
				Path:    "/work/closed/2026/07/03/TASK-086.md",
				RelPath: "closed/2026/07/03/TASK-086.md",
				Content: "## Verification\n\n- [x] ok Evidence: ../evidence/2026/07/03/TASK-086-run.md\n- [x] bad Evidence: ../evidence/2026/07/03/missing.md\n",
			},
		},
		Files: []source.File{
			{RelPath: "evidence/2026/07/03/TASK-086-run.md"},
			// missing.md is intentionally absent
		},
	}
	result := CheckEvidenceIntegrity(snapshot)
	if result.Total != 2 || result.Valid != 1 || len(result.Broken) != 1 {
		t.Fatalf("evidence = %+v, want 2 total / 1 valid / 1 broken", result)
	}
	if result.Status != StatusWarn {
		t.Errorf("status = %s, want WARN", result.Status)
	}

	// A link containing ".." is rejected and not counted.
	escSnapshot := source.Snapshot{
		ClosedFiles: []source.File{
			{
				Path:    "/work/closed/2026/07/03/TASK-001.md",
				RelPath: "closed/2026/07/03/TASK-001.md",
				Content: "Evidence: ../evidence/../../../etc/passwd.md\n",
			},
		},
	}
	escResult := CheckEvidenceIntegrity(escSnapshot)
	if escResult.Total != 0 {
		t.Errorf("escaping link counted: %+v", escResult)
	}
}

// TestVerificationClarity tests the snapshot-based clarity check without
// touching the filesystem (non-repository path).
func TestVerificationClarity(t *testing.T) {
	closedFiles := []source.File{
		{
			Path:    "/work/closed/2026/07/03/TASK-086.md",
			RelPath: "closed/2026/07/03/TASK-086.md",
			Content: `## Verification (filled at close)

- [x] proven claim (Evidence: test output)
- [x] ambiguous claim with no marker
- [ ] explicitly unproven claim
- [x] Research evidence for: skipped meta-claim

## Next Section
`,
		},
	}
	result := CheckVerificationClarity(closedFiles)
	if result.Total != 3 {
		t.Fatalf("total claims = %d, want 3 (meta-claim excluded)", result.Total)
	}
	if result.Proven != 1 || len(result.Ambiguous) != 1 || len(result.Unproven) != 1 {
		t.Fatalf("clarity = proven %d / ambiguous %d / unproven %d, want 1/1/1",
			result.Proven, len(result.Ambiguous), len(result.Unproven))
	}
	if result.Status != StatusWarn {
		t.Errorf("status = %s, want WARN", result.Status)
	}
}

func TestDeadLinks(t *testing.T) {
	root := t.TempDir()
	backlog := filepath.Join(root, "BACKLOG.md")
	plan := filepath.Join(root, "active", "TASK-001.md")
	result := CheckDeadLinks(source.Snapshot{
		Files: []source.File{
			{Path: backlog, RelPath: "BACKLOG.md", Links: []source.Link{{Href: "active/TASK-001.md"}, {Href: "active/missing.md"}}},
			{Path: plan, RelPath: "active/TASK-001.md"},
		},
		ExistingPaths: map[string]struct{}{backlog: {}, plan: {}},
	})
	if result.Scanned != 2 {
		t.Fatalf("scanned = %d, want 2", result.Scanned)
	}
	if len(result.Broken) != 1 || result.Broken[0].Link != "active/missing.md" {
		t.Fatalf("broken = %+v, want only active/missing.md (fenced link ignored)", result.Broken)
	}
	if result.Status != StatusFail {
		t.Errorf("status = %s, want FAIL", result.Status)
	}
}
